src/viam/ur/module/ur_arm_state_controlled.cpp

```diff
@@ -62,56 +62,111 @@ std::optional<URArm::state_::event_variant_> URArm::state_::state_controlled_::h
+                        if (!arm_conn_->driver->writeTrajectorySplinePoint(to_send[i].p, to_send[i].v, to_send[i].timestep)) {
+                            VIAM_SDK_LOG(error) << "send_trajectory: spline point failed; dropping connection";
+                            std::exchange(state.move_request_, {})->complete_error("failed to send trajectory spline point");
+                            return event_connection_lost_::trajectory_control_failure();
+                        }
+                    }
+
+                    mr.move_command = std::vector<trajectory_sample_point>{};
+                    return std::nullopt;
+
+                } else if (cmd.empty() && mr.cancellation_request && !mr.cancellation_request->issued) {
+                    // We have a move request, the samples have been forwarded,
+                    // and cancellation is requested but has not yet been issued. Issue a cancel.
+                    mr.cancellation_request->issued = true;
+                    if (!arm_conn_->driver->writeTrajectoryControlMessage(
+                            urcl::control::TrajectoryControlMessage::TRAJECTORY_CANCEL, 0, RobotReceiveTimeout::off())) {
+                        mr.cancel_error("failed to send trajectory cancel");
+                        VIAM_SDK_LOG(error) << "cancel failed; dropping connection";
+                        return event_connection_lost_::trajectory_control_failure();
+                    }
+                    return std::nullopt;
+                } else if (!cmd.empty() && mr.cancellation_request) {
+                    // We have a move request that we haven't issued but a
+                    // cancel is already pending. Don't issue it, just cancel it.
+                    std::exchange(state.move_request_, {})->complete_cancelled();
+                    return std::nullopt;
+
+                } else {
+                    // TODO: is it assured that we have positions/velocities here?
+                    mr.write_joint_data(state.ephemeral_->joint_positions, state.ephemeral_->joint_velocities);
+                    return std::nullopt;
+                }
+
+            } else if constexpr (std::is_same_v<T, std::optional<pose_sample>>) {
+                // Operating in pose-space
+
+                if (cmd.has_value() && !mr.cancellation_request) {
+                    // Send the single pose command
+                    if (!arm_conn_->driver->writeTrajectoryControlMessage(
+                            urcl::control::TrajectoryControlMessage::TRAJECTORY_START, 1, RobotReceiveTimeout::off())) {
+                        VIAM_SDK_LOG(error) << "single pose: start failed; dropping connection";
+                        std::exchange(state.move_request_, {})->complete_error("failed to send trajectory start for single pose");
+                        return event_connection_lost_::trajectory_control_failure();
+                    }
+
+                    auto ps = std::move(*cmd);
```

Comment:
Above, for joint space, we move the samples out right away, not after sending. Otherwise, we risk leaving the samples there despite some sort of error, and then we will try to issue it again.

I'd suggest that instead of moving out of *cmd, which moves the data out of the entry contained in the optional, but does not disengage the optional, do the same sort of thing we do above:

```
auto ps = std::exchange(cmd, {});
```

Or

```
auto ps = std::exchange(cmd, std::nullopt);
```

If you would like to be more explicit.

Then you don't need the `reset` at all, below.